### ISSUE TITLE
PDF with chinese letters is garbled

### ISSUE THREAD

**Issue by xiejx618** [asciidoctor/asciidoctor-intellij-plugin#376]
For Chinese, generating html is normal, but generating pdf is garbled

---content-----
= 标题
:lang: zh_CN
:toc:
:toclevels: 5
:sectnums:
:sectnumlevels: 5

== github使用svn下载
/tree/master/ 换成 /trunk/
/tree/${branchname}/           /branches/${branchname}/  也即将tree改成branches
---content-----

![QQ图片20200104103534](https://user-images.githubusercontent.com/2609827/71758693-0e8d9a80-2ede-11ea-9164-1c09569a851e.png)

AsciiDoc Plugin Version: 
0.30.21

IDE Name and Version: 
WebStorm 2019.3.1
Build #WS-193.5662.54, built on December 18, 2019

**Comment by ahus1** [comment]
I suppose this is a fonts problem: the standard fonts don't contain Chinese characters.

TL;DR: you might need to add some configuration with fonts to make it work.

There used to be a project here, but it is unmaintained: https://github.com/chloerei/asciidoctor-pdf-cjk-kai_gen_gothic. It contains a [sample theme](https://github.com/chloerei/asciidoctor-pdf-cjk-kai_gen_gothic/blob/master/data/themes/KaiGenGothicCN-theme.yml), but the configuration of newer versions of asciidoctor-pdf might need a little tweaking.

There is an issue that names some newer fonts: https://github.com/asciidoctor/asciidoctor-pdf/issues/1048.

I'd be grateful if you could give it a try. A small write-up of the steps needed to make it work would be a help to others. Feel free to comment this in this ticket. We could later create a wiki entry from it.

Is that something that works for you?

**Comment by ahus1** [comment]
I'm closing this for now as I haven't received feedback within a month. 
If you have feedback on this issue, please feel free to comment, and I'll re-open the issue.

**Comment by xiejx618** [comment]
webstorm 2019.3.3 + AsciiDoc 0.30.50

1.copy msyh.ttf,msyhbd-Bold.ttf(微软雅黑) fonts from windows 7.

2.add them to %USERPROFILE%\.WebStorm2019.3\config\plugins\asciidoctor-intellij-plugin\lib\asciidoctorj-pdf-1.5.3.jar\gems\asciidoctor-pdf-1.5.3\data\fonts\

3.edit %USERPROFILE%\.WebStorm2019.3\config\plugins\asciidoctor-intellij-plugin\lib\asciidoctorj-pdf-1.5.3.jar\gems\asciidoctor-pdf-1.5.3\data\themes\
----
font:
  catalog:
    #msyh supports simplified chinese
    msyh:
      normal: msyh.ttf
      bold: msyhbd-Bold.ttf
      italic: msyh.ttf
      bold_italic: msyhbd-Bold.ttf

#omit the other
base:
  font_family: msyh
----
the whole content:
----
font:
  catalog:
    #msyh supports simplified chinese
    msyh:
      normal: msyh.ttf
      bold: msyhbd-Bold.ttf
      italic: msyh.ttf
      bold_italic: msyhbd-Bold.ttf
    # Noto Serif supports Latin, Latin-1 Supplement, Latin Extended-A, Greek, Cyrillic, Vietnamese & an assortment of symbols
    Noto Serif:
      normal: notoserif-regular-subset.ttf
      bold: notoserif-bold-subset.ttf
      italic: notoserif-italic-subset.ttf
      bold_italic: notoserif-bold_italic-subset.ttf
    # M+ 1mn supports ASCII and the circled numbers used for conums
    M+ 1mn:
      normal: mplus1mn-regular-subset.ttf
      bold: mplus1mn-bold-subset.ttf
      italic: mplus1mn-italic-subset.ttf
      bold_italic: mplus1mn-bold_italic-subset.ttf
page:
  background_color: FFFFFF
  layout: portrait
  initial_zoom: FitH
  margin: [0.5in, 0.67in, 0.67in, 0.67in]
  # margin_inner and margin_outer keys are used for recto/verso print margins when media=prepress
  margin_inner: 0.75in
  margin_outer: 0.59in
  size: A4
base:
  align: justify
  # color as hex string (leading # is optional)
  font_color: 333333
  # color as RGB array
  #font_color: [51, 51, 51]
  # color as CMYK array (approximated)
  #font_color: [0, 0, 0, 0.92]
  #font_color: [0, 0, 0, 92%]
  font_family: msyh
  # choose one of these font_size/line_height_length combinations
  #font_size: 14
  #line_height_length: 20
  #font_size: 11.25
  #line_height_length: 18
  #font_size: 11.2
  #line_height_length: 16
  font_size: 10.5
  #line_height_length: 15
  # correct line height for Noto Serif metrics
  line_height_length: 12
  #font_size: 11.25
  #line_height_length: 18
  line_height: $base_line_height_length / $base_font_size
  font_size_large: round($base_font_size * 1.25)
  font_size_small: round($base_font_size * 0.85)
  font_size_min: $base_font_size * 0.75
  font_style: normal
  border_color: EEEEEE
  border_radius: 4
  border_width: 0.5
role:
  line-through:
    text_decoration: line-through
  underline:
    text_decoration: underline
  big:
    font_size: $base_font_size_large
  small:
    font_size: $base_font_size_small
  subtitle:
    font_size: 0.8em
    font_color: 999999
# FIXME vertical_rhythm is weird; we should think in terms of ems
#vertical_rhythm: $base_line_height_length * 2 / 3
# correct line height for Noto Serif metrics (comes with built-in line height)
vertical_rhythm: $base_line_height_length
horizontal_rhythm: $base_line_height_length
# QUESTION should vertical_spacing be block_spacing instead?
vertical_spacing: $vertical_rhythm
link:
  font_color: 428BCA
# literal is currently used for inline monospaced in prose and table cells
literal:
  font_color: B12146
  font_family: M+ 1mn
button:
  content: "[\u2009%s\u2009]"
  font_style: bold
key:
  background_color: F5F5F5
  border_color: CCCCCC
  border_offset: 2
  border_radius: 2
  border_width: 0.5
  font_family: $literal_font_family
  separator: "\u202f+\u202f"
mark:
  background_color: FFFF00
  border_offset: 1
menu:
  caret_content: " <font size=\"1.15em\"><color rgb=\"b12146\">\u203a</color></font> "
heading:
  align: left
  font_color: $base_font_color
  font_style: bold
  # h1 is used for part titles (book doctype) or the doctitle (article doctype)
  h1_font_size: floor($base_font_size * 2.6)
  # h2 is used for chapter titles (book doctype only)
  h2_font_size: floor($base_font_size * 2.15)
  h3_font_size: round($base_font_size * 1.7)
  h4_font_size: $base_font_size_large
  h5_font_size: $base_font_size
  h6_font_size: $base_font_size_small
  #line_height: 1.4
  # correct line height for Noto Serif metrics (comes with built-in line height)
  line_height: 1
  margin_top: $vertical_rhythm * 0.4
  margin_bottom: $vertical_rhythm * 0.9
  min_height_after: $base_line_height_length * 1.5
title_page:
  align: right
  logo:
    top: 10%
  title:
    top: 55%
    font_size: $heading_h1_font_size
    font_color: 999999
    line_height: 0.9
  subtitle:
    font_size: $heading_h3_font_size
    font_style: bold_italic
    line_height: 1
  authors:
    margin_top: $base_font_size * 1.25
    font_size: $base_font_size_large
    font_color: 181818
  revision:
    margin_top: $base_font_size * 1.25
block:
  margin_top: 0
  margin_bottom: $vertical_rhythm
caption:
  align: left
  font_size: $base_font_size * 0.95
  font_style: italic
  # FIXME perhaps set line_height instead of / in addition to margins?
  margin_inside: $vertical_rhythm / 3
  #margin_inside: $vertical_rhythm / 4
  margin_outside: 0
lead:
  font_size: $base_font_size_large
  line_height: 1.4
abstract:
  font_color: 5C6266
  font_size: $lead_font_size
  line_height: $lead_line_height
  font_style: italic
  first_line_font_style: bold
  title:
    align: center
    font_color: $heading_font_color
    font_size: $heading_h4_font_size
    font_style: $heading_font_style
admonition:
  column_rule_color: $base_border_color
  column_rule_width: $base_border_width
  padding: [0, $horizontal_rhythm, 0, $horizontal_rhythm]
  #icon:
  #  tip:
  #    name: far-lightbulb
  #    stroke_color: 111111
  #    size: 24
  label:
    text_transform: uppercase
    font_style: bold
blockquote:
  font_size: $base_font_size_large
  border_color: $base_border_color
  border_width: 0
  border_left_width: 5
  # FIXME disable negative padding bottom once margin collapsing is implemented
  padding: [0, $horizontal_rhythm, $block_margin_bottom * -0.75, $horizontal_rhythm + $blockquote_border_left_width / 2]
  cite_font_size: $base_font_size_small
  cite_font_color: 999999
verse:
  font_size: $blockquote_font_size
  border_color: $blockquote_border_color
  border_width: $blockquote_border_width
  border_left_width: $blockquote_border_left_width
  padding: $blockquote_padding
  cite_font_size: $blockquote_cite_font_size
  cite_font_color: $blockquote_cite_font_color
# code is used for source blocks (perhaps change to source or listing?)
code:
  font_color: $base_font_color
  font_family: $literal_font_family
  font_size: ceil($base_font_size)
  padding: $code_font_size
  line_height: 1.25
  # line_gap is an experimental property to control how a background color is applied to an inline block element
  line_gap: 3.8
  background_color: F5F5F5
  border_color: CCCCCC
  border_radius: $base_border_radius
  border_width: 0.75
conum:
  font_family: $literal_font_family
  font_color: $literal_font_color
  font_size: $base_font_size
  line_height: 4 / 3
  glyphs: circled
example:
  border_color: $base_border_color
  border_radius: $base_border_radius
  border_width: 0.75
  background_color: $page_background_color
  # FIXME reenable padding bottom once margin collapsing is implemented
  padding: [$vertical_rhythm, $horizontal_rhythm, 0, $horizontal_rhythm]
image:
  align: left
prose:
  margin_top: $block_margin_top
  margin_bottom: $block_margin_bottom
sidebar:
  background_color: EEEEEE
  border_color: E1E1E1
  border_radius: $base_border_radius
  border_width: $base_border_width
  # FIXME reenable padding bottom once margin collapsing is implemented
  padding: [$vertical_rhythm, $vertical_rhythm * 1.25, 0, $vertical_rhythm * 1.25]
  title:
    align: center
    font_color: $heading_font_color
    font_size: $heading_h4_font_size
    font_style: $heading_font_style
thematic_break:
  border_color: $base_border_color
  border_style: solid
  border_width: $base_border_width
  margin_top: $vertical_rhythm * 0.5
  margin_bottom: $vertical_rhythm * 1.5
description_list:
  term_font_style: bold
  term_spacing: $vertical_rhythm / 4
  description_indent: $horizontal_rhythm * 1.25
outline_list:
  indent: $horizontal_rhythm * 1.5
  #marker_font_color: 404040
  # NOTE outline_list_item_spacing applies to list items that do not have complex content
  item_spacing: $vertical_rhythm / 2
table:
  background_color: $page_background_color
  border_color: DDDDDD
  border_width: $base_border_width
  cell_padding: 3
  head:
    font_style: bold
    border_bottom_width: $base_border_width * 2.5
  body:
    stripe_background_color: F9F9F9
  foot:
    background_color: F0F0F0
toc:
  indent: $horizontal_rhythm
  line_height: 1.4
  dot_leader:
    #content: ". "
    font_color: A9A9A9
    #levels: 2 3
footnotes:
  font_size: round($base_font_size * 0.75)
  item_spacing: $outline_list_item_spacing / 2
header:
  font_size: $base_font_size_small
  line_height: 1
  vertical_align: middle
footer:
  font_size: $base_font_size_small
  # NOTE if background_color is set, background and border will span width of page
  border_color: DDDDDD
  border_width: 0.25
  height: $base_line_height_length * 2.5
  line_height: 1
  padding: [$base_line_height_length / 2, 1, 0, 1]
  vertical_align: top
  recto:
    #columns: "<50% =0% >50%"
    right:
      content: '{page-number}'
  verso:
    #columns: $footer_recto_columns
    left:
      content: $footer_recto_right_content
----

**Comment by xiejx618** [comment]
![image](https://user-images.githubusercontent.com/2609827/78211157-cbdc5e00-74de-11ea-9906-201720298eff.png)

**Comment by ahus1** [comment]
Thank you for sharing this! I'll give it a try and put it into the plugin's GitHub wiki.

When adding a theme to your project, it shouldn't be necessary to change the default theme distributed by the plugin: you can add 

```
:pdf-stylesdir: (folder local to doc)
```

to the document. See https://github.com/ahus1/asciidoctor-deepdive/blob/master/manual/src/main/asciidoc/manual.adoc for an example.

**Comment by mojavelinux** [comment]
I agree with @ahus1, you should not be modifying the gem. That's just going to get you into trouble. You also don't have to redeclare all the properties in the default theme. Instead, you can extend the theme instead. See https://github.com/asciidoctor/asciidoctor-pdf/blob/master/docs/theming-guide.adoc#keys-extends

**Comment by ahus1** [comment]
@xiejx618 - thank you for the example. I used it to build a small example that doesn't need configurations or fonts to be added to JAR files. You'll find a screenshot below of some random text.
Unfortunately I'm not familiar with your language, therefore I can't verify the result. 

You can find it here: https://github.com/ahus1/asciidoctor-deepdive/tree/master/manual-zh_CN

The following should work: 
* check out the repository
* follow the instructions in https://github.com/ahus1/asciidoctor-deepdive/blob/master/manual-zh_CN/css/zh_CN-theme.yml to select the font you want to use. 
* download the font files and copy them to the fonts directory
* open the file `manual.adoc` in your IDE with the plugin
* create a PDF using the button in the IDE **or** run asciidoctor-pdf from the command 

Both IDE and the command line should generate the same result. 

Please let me know if this is missing anything. 

@diguage - I know you've been into Asciidoctor and PDF generation as well. Maybe you have suggestions how to improve.  

![image](https://user-images.githubusercontent.com/3957921/78456708-8ae27600-76a5-11ea-9cda-1ed93c3e6207.png)

**Comment by mojavelinux** [comment]
Note that you can now merge the font catalog with the inherited one, so there's no need to redeclare Noto Serif, M+ 1mn, and M+ 1p Fallback. See https://github.com/asciidoctor/asciidoctor-pdf/blob/master/docs/theming-guide.adoc#extends

**Comment by diguage** [comment]
https://github.com/diguage/jdk-source-analysis

The repo use Maven plugin to generate PDF. @xiejx618  This might be useful for you. 

I use Source Han Serif SC font. You can use https://raw.githubusercontent.com/diguage/jdk-source-analysis/master/cfg/fonts/download-fonts.sh Bash Shell script to download the fonts.

**Comment by xiejx618** [comment]
> @xiejx618 - thank you for the example. I used it to build a small example that doesn't need configurations or fonts to be added to JAR files. You'll find a screenshot below of some random text.
> Unfortunately I'm not familiar with your language, therefore I can't verify the result.
> 
> You can find it here: https://github.com/ahus1/asciidoctor-deepdive/tree/master/manual-zh_CN
> 
> The following should work:
> 
> * check out the repository
> * follow the instructions in https://github.com/ahus1/asciidoctor-deepdive/blob/master/manual-zh_CN/css/zh_CN-theme.yml to select the font you want to use.
> * download the font files and copy them to the fonts directory
> * open the file `manual.adoc` in your IDE with the plugin
> * create a PDF using the button in the IDE **or** run asciidoctor-pdf from the command
> 
> Both IDE and the command line should generate the same result.
> 
> Please let me know if this is missing anything.
> 
> @diguage - I know you've been into Asciidoctor and PDF generation as well. Maybe you have suggestions how to improve.
> 
> ![image](https://user-images.githubusercontent.com/3957921/78456708-8ae27600-76a5-11ea-9cda-1ed93c3e6207.png)

yes，it works
